lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -4028,6 +4028,26 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     if (toType->isAnyObject() || fromType->isAnyObject())
       return CheckedCastKind::ValueCast;
 
+    // A checked cast from Swift.Error to a type that does not conform to
+    // Swift.Error cannot succeed.
+    //
+    // struct S {}
+    // enum FooError: Error { case bar }
+    // func foo() throws {
+    //   do {
+    //     throw FooError.bar
+    //   } catch is X {
+    //     print("Caught bar error")
+    //   }
+    // }
+    auto exceptionType = getExceptionType(dc, SourceLoc());
+    auto toTypeConformsToError =
+        conformsToProtocol(toType, Context.getErrorDecl(), dc,
+                           ConformanceCheckFlags::InExpression);
+    if (fromType->isEqual(exceptionType) && !toTypeConformsToError) {
```

Comment:
Nitpick: checking protocol conformance isn't cheap, so please delay that until after you've checked that we're casting from `Error`.

Is this true of any protocol, or just Error?

What happens in a case like this?

```swift
func test1(error: Error) {
  _ = error as? (NSObject & Error)
}
func test2(error: Error) {
  _ = error as? (Error & Comparable)
}
```

Reply:
@jrose-apple Done.

It checks conformance to `Error`, so these would pass the check:

```swift
enum Foo: Error { case bar }

class X: NSObject, Error {}
class Y: NSError {}
class Z {}

// Ok
do {
  throw Foo.bar
} catch is X {
  print("Foo")
}

// Ok
do {
  throw Foo.bar
} catch is Y {
  print("Foo")
}

// Warning
do {
  throw Foo.bar
} catch is Z {
  print("Foo")
}
```

Reply:
No, I mean literally `catch is (NSObject & Error)` and `catch is (Error & Comparable)`.

Reply:
Your first example does not hit this new logic. Second one is not valid as Comparable has constraints.

Reply:
Or

```swift
func test3<T>(error: Error, as type: T.Type) -> Bool {
  return (error as? T) != nil
}
```

I don't think this one works either.

Reply:
>< Sorry. `catch is (NSViewController & Error)` and `catch is (Error & NSApplicationDelegate)`, then.

Reply:
More examples:

```swift
class A {}
class B: Error & NSObject {}

// Warning
do {
  throw Foo.bar
} catch let error as A {
  print("Foo")
}

// Ok
do {
  throw Foo.bar
} catch let error as B {
  print("Foo")
}
```

Reply:
@jrose-apple Oh, I see. Hmm `cast from 'Error' to unrelated type 'NSApplicationDelegate & Error' always fails`

Reply:
I guess it makes sense, but if we want to check if the protocol composition type contains `Error` then we can use `containsProtocol()` instead of checking for conformance and I guess it would work in general as well (unless I am missing something).

Your example,
```
func test3<T>(error: Error, as type: T.Type) -> Bool {
  return (error as? T) != nil
}
```

leads to the cast diagnostic as expected.

Reply:
That one should *not* lead to the cast diagnostic; it is possible `T` is dynamically a type that conforms to Error.

Reply:
I think the conclusion I'm coming to is that there's some general "this cast can't succeed" check that we might be missing, but this isn't it. It's possible it's because we wanted to handle someone adding a retroactive conformance in another module, but that seems like a code smell.

Maybe it'd be restrictive enough to say "am I casting from a protocol (existential) type to a concrete type", which is easy to check with `getAnyNominal()`.

Reply:
Ah right, we'd also have to check if `toType` is an archetype type or not for that to not get diagnosed.

Do you mean casting from an existential type to a type that does not conform to that type? That would be a more general check.

Reply:
The `T` case is an archetype. `NSApplicationDelegate & Error` is a protocol composition. What other cases are we not thinking about?

I picked a case we know it'd be reasonable to warn on: you're starting from a protocol type and you're casting to a concrete type. That's more general than Error-to-concrete, sure, but Error isn't really special here. The important part is restricting the `toType`.

Reply:
Hmm okay. I've tweaked the code to make it more general!